### DISCUSSION TITLE
Generating Skybox KTX files

### DISCUSSION THREAD

**Discussion by omri-polycam** [google/filament#6736] [General]
Hi guys, I'm Omri, I'm the main rendering engineering working with Chris at Polycam.
I am currently trying to integrate compressed textures for use with skyboxes to allow for higher resolution cubemap images on texture memory limited mobile devices. We also plan to use this format to take advantage of preprocessed mipmaps.
I'm currently a bit stuck trying to find the best way to generate the .ktx files to load into `ktxreader::Ktx1Reader` to feed into `filament::Skybox::Builder::environment()`. I'd appreciate any advice on what's the best way to generate a filament compatible .ktx file from either a single panorama png/jpg or 6 cube map side png/jpg images. Below are some of the approaches I've tried:

- `ktreader::Ktx2Reader` with .ktx2 files generated from [basisu](https://github.com/BinomialLLC/basis_universal). The problem is that these don't work because I hit an error in `ktreader::Ktx2Reader` saying that cubemaps are not yet supported for .ktx2 files.
- Using the `cmgen` tool built into filament. This worked, but these take an extraordinarily long time to process, presumably generating the spherical harmonics and other lighting information that I don't even need. Our app is unfortunately not a typical filament user like building a game or such where all the assets are preprocessed into the executable. We will be generating new .ktx files from our user's scans on a regular basis, so acceptable compression performance is a requirement. Is there a way to use this tool to just generate the compressed cubemap .ktx texture and that's it?
- Using [basisu](https://github.com/BinomialLLC/basis_universal) to first generate a .basis file and transcode that into a .ktx. This works but the colors end up quite a bit lighter than the original 6 cubemap images. I have tried various sRGB and linear related command line options into `basisu` to no effect. I've also experimented with the final `srgb` boolean argument into `ktxreader::Ktx1Reader::createTexture()` and that has not fixed the problem either. Setting this argument to true still succeeds but warns that the image is linear and not sRGB. As best as I can tell `basisu` is messing up the colors in the final .basis to .ktx transcoding step. If anybody at filament has any experience with this `basisu` command line tool, do you know what options to pass to maintain color fidelity? Below are the pertinent lines in the script:
```
basisu -tex_type cubemap Face_2.jpg Face_0.jpg Face_4.jpg Face_5.jpg Face_1.jpg Face_3.jpg
basisu -unpack -ktx_only -format_only 1 Face_2.basis
```
- Using Khronos' [toktx](https://github.com/KhronosGroup/KTX-Software) command line tool. This unfortunately generates the error `utils::PreconditionPanic reason: KTX has unexpected identifier` whenever I use the `--encode` option into `toktx` when calling `ktxreader::Ktx1Reader::createTexture()`. Without the `--encode` option, it does load fine but then the .ktx file contains no compression so this defeats the point.

If anybody knows of a different approach or has any advice on tweaking any of the approaches I've already tried to work correctly, I'd greatly appreciate it.

Thanks,
   Omri

**Comment by romainguy** [comment]
cmgen can extract only a skybox file without doing all the integration for lighting/ibl. I don't remember the flags off the top of my head but it's possible (I often use this to generate higher resolution skyboxes for my own tests).

**Reply by omri-polycam** [reply]
Thanks, so I got it to generate .ktx files only, but they don't seem to be compressed which is my main motivation for using .ktx files. What's passed into `Texture::setImage()` inside `ktxreader::Ktx1Reader::createTexture()` for the size parameter of the `PixelBufferDescriptor` is 100,663,296 bytes. This is precisely 2,048 x 2,048 x 4 x 6 (width x height x bytes per pixel x cube faces). Also the `isCompressed(ktxinfo)` check in `ktxreader::Ktx1Reader::createTexture()` returns `false`. According to the cmgen help, the `--compression` argument is ignored for .ktx files and they are encoded with `RGB_10_11_11_REV` data. Is there a way to get cmgen to created compressed ktx files or do I need to use another tool or another format? Thanks!

**Reply by pannonianknight** [reply]
have u sorted this out?
